sw/device/lib/crypto/impl/kdf.c

```diff
@@ -15,13 +15,179 @@
     otcrypto_blinded_key_t *keying_material) {
-  // TODO: Implement HMAC-KDF-CTR.
-  return OTCRYPTO_NOT_IMPLEMENTED;
+  // Check NULL pointers.
+  if (keying_material == NULL || keying_material->keyblob == NULL ||
+      key_derivation_key.keyblob == NULL) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  if (launder32(keying_material->config.security_level) !=
+          kOtcryptoKeySecurityLevelLow ||
+      launder32(key_derivation_key.config.security_level) !=
+          kOtcryptoKeySecurityLevelLow) {
+    // The underlying HMAC implementation is not currently hardened.
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  }
+
+  // Check for null label with nonzero length.
+  if (kdf_label.data == NULL && kdf_label.len != 0) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check for null context with nonzero length.
+  if (kdf_context.data == NULL && kdf_context.len != 0) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check the private key checksum.
+  if (integrity_blinded_key_check(&key_derivation_key) != kHardenedBoolTrue) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check non-zero length for keying_material.
+  if (required_byte_len == 0) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Infer the digest size.
+  size_t digest_word_len = 0;
+  HARDENED_TRY(digest_num_words_from_key_mode(
+      key_derivation_key.config.key_mode, &digest_word_len));
+
+  // Ensure that the derived key is a symmetric key masked with XOR and is not
+  // supposed to be hardware-backed.
+  HARDENED_TRY(keyblob_ensure_xor_masked(keying_material->config));
+
+  // Check `keying_material` key length.
+  if (keying_material->config.hw_backed == kHardenedBoolTrue) {
+    // The case where `keying_material` is hw_backed is addressed by
+    // `otcrypto_hw_backed_key` function in `key_transport.h`.
+    return OTCRYPTO_BAD_ARGS;
+  } else if (keying_material->config.hw_backed == kHardenedBoolFalse) {
+    if (keying_material->config.key_length != required_byte_len ||
+        keying_material->keyblob_length !=
+            keyblob_num_words(keying_material->config) * sizeof(uint32_t)) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+  } else {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check that the unmasked key length is not too large for HMAC CTR
+  // (see NIST SP 800-108r1, section 4.1)
+  size_t required_word_len = ceil_div(required_byte_len, sizeof(uint32_t));
+  size_t num_iterations = ceil_div(required_word_len, digest_word_len);
+  if (launder32(num_iterations) > UINT32_MAX) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+  HARDENED_CHECK_LE(num_iterations, UINT32_MAX);
+
+  // Check if label or context contain 0x00 bytes
+  // Since 0x00 is used as the delimiter between label and context
+  // there shouldn't be multiple instances of them in input data
+  HARDENED_TRY(check_zero_byte(kdf_label));
+  HARDENED_TRY(check_zero_byte(kdf_context));
+
+  // Setup
+  uint8_t zero = 0x00;
+  uint32_t required_bit_len = __builtin_bswap32(required_byte_len * 8);
```

Comment:
nit: should maybe check that `required_byte_len * 8` doesn't overflow a `uint32_t`.